klog/klog.c

```diff
@@ -0,0 +1,327 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system kernel
+ *
+ * Kernel log buffer
+ *
+ * Copyright 2021 Phoenix Systems
+ * Author: Maciej Purski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include HAL
+
+#include "../include/ioctl.h"
+
+#include "../posix/posix.h"
+
+#include "klog.h"
+#include "../proc/threads.h"
+#include "../proc/msg.h"
+#include "../proc/ports.h"
+
+
+#ifndef KLOG_BUFSZ
+#define KLOG_BUFSZ (2 * SIZE_PAGE)
+#endif
+
+
+#define TCGETS 0x405c7401
+
+
+typedef struct klog_reader {
+	msg_t msg;
+	offs_t ridx;
+	unsigned long rid;
+	unsigned pid;
+	unsigned nonblocking;
+	unsigned waiting;
+	struct klog_reader *prev, *next;
+} klog_reader_t;
+
+
+static struct {
+	char buf[KLOG_BUFSZ];
+	u32 port;
+	offs_t head;
+	offs_t tail;
+	lock_t lock;
+	klog_reader_t *readers;
+} klog_common;
+
+
+static int _fifo_empty(void)
+{
+	return klog_common.tail == klog_common.head;
+}
+
+
+static int _fifo_full(void)
+{
+	return klog_common.tail - klog_common.head == KLOG_BUFSZ;
+}
+
+
+static char _fifo_pop(void)
+{
+	return klog_common.buf[klog_common.head++ % KLOG_BUFSZ];
+}
+
+
+static void _fifo_push(char c)
+{
+	klog_common.buf[klog_common.tail % KLOG_BUFSZ] = c;
+	klog_common.tail++;
+	if (_fifo_full())
+		_fifo_pop();
+}
+
+
+static char _fifo_get(offs_t off)
+{
+	return klog_common.buf[off % KLOG_BUFSZ];
+}
+
+
+void _klog_init(void)
+{
+	hal_memset(&klog_common, 0, sizeof(klog_common));
+	proc_lockInit(&klog_common.lock);
+}
+
+
+static klog_reader_t *klog_readerFind(unsigned pid)
+{
+	klog_reader_t *r = klog_common.readers;
+
+	if (r != NULL) {
+		do {
+			if (r->pid == pid)
+				return r;
+			r = r->next;
+		} while (r != klog_common.readers);
+	}
+
+	return NULL;
+}
+
+
+static int klog_readerAdd(unsigned pid, unsigned nonblocking)
+{
+	klog_reader_t *r;
+
+	if (klog_readerFind(pid) != NULL)
+		return -EINVAL;
+
+	if ((r = vm_kmalloc(sizeof(klog_reader_t))) == NULL)
+		return -ENOMEM;
+
+	r->next = NULL;
+	r->prev = NULL;
+	r->nonblocking = !!nonblocking;
+	r->pid = pid;
+	r->waiting = 0;
+	hal_memset(&r->msg, 0, sizeof(msg_t));
+
+	proc_lockSet(&klog_common.lock);
+	r->ridx = klog_common.head;
+	LIST_ADD(&klog_common.readers, r);
+	proc_lockClear(&klog_common.lock);
+
+	return 0;
+}
+
+
+static int _klog_readln(klog_reader_t *r, char *buf, size_t sz)
+{
+	size_t n = 0;
+	char c;
+
+	while (r->ridx < klog_common.tail && n < sz) {
+		c = _fifo_get(r->ridx++);
+		buf[n++] = c;
+		if (c == '\n' || c == '\0')
+			break;
```

Comment:
We don't really need c
```suggestion
		buf[n++] = _fifo_get(r->ridx++);
		if (buf[n] == '\n' || buf[n] == '\0')
			break;
```

Reply:
I agree, but the code you have suggested is not an equivalent of what I have written. There should be `buf[n - 1]`.

Reply:
Right, sorry